### PR TITLE
swap adaptor changes

### PR DESCRIPTION
**Pull request by divyaN73**

### CODE REVIEW

File: contracts/hyphen/swaps/UniswapAdaptor.sol [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+import "../interfaces/ISwapRouter.sol";
+import "../lib/TransferHelper.sol";
+
+pragma abicoder v2;
+
+contract UniswapAdaptor is ISwapAdaptor {
+
+    ISwapRouter public immutable swapRouter;
+
+    // For this example, we will set the pool fee to 0.3%.
```
Comment (tomarsachin2271) [inline contracts/hyphen/swaps/UniswapAdaptor.sol:18]: remove this comment.

File: contracts/hyphen/interfaces/ISwapAdaptor.sol [hunk 2]
```diff
@@ -0,0 +1,12 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.0;
+import "../structures/SwapRequest.sol";
+
+interface ISwapAdaptor {
+    function swap(
```
[the remark this answers is not in the document]
Comment (divyaN73) [reply contracts/hyphen/interfaces/ISwapAdaptor.sol:6]: no need, as discussed

File: contracts/hyphen/swaps/UniswapAdaptor.sol [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    /// @notice swapExactOutputSingle swaps a minimum possible amount of DAI for a fixed amount of WETH.
+    /// @dev The calling address must approve this contract to spend its DAI for this function to succeed. As the amount of input DAI is variable,
+    /// the calling address will need to approve for a slightly higher amount, anticipating some variance.
+    /// @param amountInMaximum The amount of DAI we are willing to spend to receive the specified amount of WETH9.
+    /// @return amountIn The amount of DAI actually spent in the swap.
+    function swap(
+        address inputTokenAddress,
+        uint256 amountInMaximum,
+        address receiver,
+        SwapRequest[] memory swapRequests
```
Comment (ankurdubey521) [inline contracts/hyphen/swaps/UniswapAdaptor.sol:32]: use calldata instead of memory to save gas

File: contracts/hyphen/LiquidityPool.sol [hunk 4]
```diff
@@ -180,23 +197,65 @@ contract LiquidityPool is
...
+     * @param swapRequest information related to token swap on exit chain
+     */
+    function depositAndSwapErc20(
+        address tokenAddress,
+        address receiver,
+        uint256 toChainId,
+        uint256 amount,
+        string calldata tag,
+        SwapRequest[] calldata swapRequest
+    ) public tokenChecks(tokenAddress) whenNotPaused nonReentrant {
```
Comment (ankurdubey521) [inline contracts/hyphen/LiquidityPool.sol:223]: use external instead of public to save some gas

File: contracts/hyphen/LiquidityPool.sol [hunk 5]
```diff
@@ -180,23 +197,65 @@ contract LiquidityPool is
...
+    
+        uint256 totalPercentage = 0;
+        for (uint256 index = 0; index < swapRequest.length; ++index) {
```
Comment (ankurdubey521) [inline contracts/hyphen/LiquidityPool.sol:226]: create a separate variable for swapRequest.length, and enclose the for loop inside unchecked {} to save gas

File: contracts/hyphen/LiquidityPool.sol [hunk 6]
```diff
@@ -271,6 +330,32 @@ contract LiquidityPool is
...
     ) external payable whenNotPaused nonReentrant {
+        uint256 rewardAmount = _depositNative(receiver, toChainId);
+        emit Deposit(_msgSender(), NATIVE, receiver, toChainId, msg.value + rewardAmount, rewardAmount, tag);
+    }
+
+    function depositNativeAndSwap(
+        address receiver,
+        uint256 toChainId,
+        string calldata tag,
+        SwapRequest[] memory swapRequest
```
Comment (ankurdubey521) [inline contracts/hyphen/LiquidityPool.sol:341]: use calldata instead of memory

File: contracts/hyphen/LiquidityPool.sol [hunk 7]
```diff
@@ -271,6 +330,32 @@ contract LiquidityPool is
...
+    ) external payable whenNotPaused nonReentrant {
+
+        uint256 totalPercentage = 0;
+        for (uint256 index = 0; index < swapRequest.length; ++index) {
```
Comment (ankurdubey521) [inline contracts/hyphen/LiquidityPool.sol:345]: create separate variable for swapRequest.length and wrap the whole loop in unchecked {} to save gas

File: contracts/hyphen/LiquidityPool.sol [hunk 8]
```diff
@@ -379,29 +464,50 @@ contract LiquidityPool is
...
-        liquidityProviders.decreaseCurrentLiquidity(tokenAddress, transferDetails[0]);
+    function swapAndSendFundsToUser(
+        address tokenAddress,
+        uint256 amount,
+        address payable receiver,
+        bytes calldata depositHash,
+        uint256 nativeTokenPriceInTransferredToken,
+        uint256 fromChainId,
+        uint256 swapGasOverhead,
+        SwapRequest[] memory swapRequests,
```
Comment (ankurdubey521) [inline contracts/hyphen/LiquidityPool.sol:497]: use calldata

File: contracts/hyphen/LiquidityPool.sol [hunk 9]
```diff
@@ -379,29 +464,50 @@ contract LiquidityPool is
...
+        string memory swapAdaptor
```
Comment (ankurdubey521) [inline contracts/hyphen/LiquidityPool.sol:498]: use calldata

**Review by ankurdubey521 — CHANGES_REQUESTED** [review]

**Comment by ankurdubey521** [comment]
We should also update [the Upgradibility Tests](https://github.com/bcnmy/hyphen-contract/blob/HP-60/test/Upgradibility.tests.ts) since we're upgrading the Liquidity Pool, but this not necessarily required at the moment

**Review by ankurdubey521 — APPROVED** [review]

File: contracts/hyphen/swaps/UniswapAdaptor.sol [hunk 10]
```diff
@@ -11,42 +11,114 @@ import "../lib/TransferHelper.sol";
...
-    ) override external returns (uint256 amountIn) {
+    ) override external returns (uint256 amountOut) {
+
+        require(inputTokenAddress != NATIVE, "wrong function");
+        uint256 swapArrayLength = swapRequests.length;
+
+        // Only first element should have ExactOutput operation
+        if(swapArrayLength > 1) {
+            for( uint256 index=0; index < swapArrayLength; index++ ) {
+                require(index == 0 || swapRequests[index].operation != SwapOperation.ExactOutput, "Invalid swap request");
```
Comment (tomarsachin2271) [inline contracts/hyphen/swaps/UniswapAdaptor.sol:42]: require(swapArrayLength == 1 || swapRequests[1].operation == SwapOperation.ExactInput, "Invalid swap operation")

File: contracts/hyphen/swaps/UniswapAdaptor.sol [hunk 11]
```diff
@@ -14,18 +14,20 @@ contract UniswapAdaptor is ISwapAdaptor {
     uint24 public constant POOL_FEE = 3000;
     address private constant NATIVE = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
     
-    address public immutable WRAPPED_NATIVE_TOKEN_ADDRESS;
+    address public immutable WRAPPER_ADDRESS;
```
Comment (tomarsachin2271) [inline contracts/hyphen/swaps/UniswapAdaptor.sol:17]: Change name to NATIVE_WRAPPED_ADDRESS

**Review by ankurdubey521 — APPROVED** [review]

**Pull request merged**